examples/measure/rs-measure.cpp

```diff
@@ -0,0 +1,553 @@
+
+// Distance 3D is used to calculate real 3D distance between two pixels
+float dist_3d(const rs2_intrinsics& intr, const rs2::depth_frame& frame, pixel u, pixel v);
+// Distance 2D returns the distance in pixels between two pixels
+float dist_2d(const pixel& a, const pixel& b);
+
+// Toggle helper class will be used to render the two buttons
+// controlling the edges of our ruler
+struct toggle
+{
+    toggle() : x(0.f), y(0.f) {}
+    toggle(float x, float y) 
+        : x(std::min(std::max(x, 0.f), 1.f)), 
+          y(std::min(std::max(y, 0.f), 1.f))
+    {}
+
+    // Move from [0,1] space to pixel space of specific frame
+    pixel get_pixel(rs2::depth_frame frame) const
+    {
+        int px = x * frame.get_width();
+        int py = y * frame.get_height();
+        return{ px, py };
+    }
+
+    void render(const window& app)
+    {
+        // Render a circle
+        const float r = 10;
+        const float segments = 16;
+        glEnable(GL_BLEND);
+        glBlendFunc(GL_ONE_MINUS_DST_COLOR, GL_ONE_MINUS_SRC_COLOR);
+        glColor3f(0.f, 1.0f, 0.0f);
+        glLineWidth(2);
+        glBegin(GL_LINE_STRIP);
+        for (auto i = 0; i <= segments; i++)
+        {
+            auto t = 2 * M_PI * float(i) / segments;
+            glVertex2f(x * app.width()  + cos(t) * r, 
+                       y * app.height() + sin(t) * r);
+        }
+        glEnd();
+        glDisable(GL_BLEND);
+        glColor3f(1.f, 1.f, 1.f);
+    }
+
+    // This helper function is used to find the button
+    // closest to the mouse cursor
+    // Since we are only comparing this distance, sqrt can be safely skipped 
+    float dist_2d(const toggle& other) const
+    {
+        return pow(x - other.x, 2) + pow(y - other.y, 2);
+    }
+
+    float x;
+    float y;
+    bool selected = false;
+};
+
+// Application state shared between the main-thread and GLFW events
+struct state 
+{
+    bool mouse_down = false;
+    toggle ruler_start;
+    toggle ruler_end;
+};
+
+// Helper function to register to UI events
+void register_glfw_callbacks(window& app, state& app_state);
+
+// Distance rendering functions:
+
+// Simple distance is the classic pythagorean distance between 3D points
+// This distance ignores the topology of the object and can cut both through
+// air and through solid
+void render_simple_distance(const rs2::depth_frame& depth, 
+                            const state& s, 
+                            const window& app,
+                            const rs2_intrinsics& intr);
+
+// Shortest-path distance approximates the geodesic.
+// Given two points on a surface it will follow with that surface
+void render_shortest_path(const rs2::depth_frame& depth,
+                          const std::vector<pixel>& path,
+                          const window& app,
+                          float total_dist);
+
+int main(int argc, char * argv[]) try
+{
+    // OpenGL textures for the color and depth frames
+    texture depth_image, color_image;
```

Comment:
I think a less confusing name should be used such as `*_view` or `*_texture` (in the main loop you have a depth frame and depth image)